test_result.go

```diff
@@ -0,0 +1,405 @@
+func (m buildTestResultModule) getBuildTestResultSummaryHandler(c *gin.Context) {
+	buildID, ok := ginutil.ParseParamUint(c, "buildid")
+	if !ok {
+		return
+	}
+
+	artifactID, ok := ginutil.ParseParamUint(c, "artifactId")
+	if !ok {
+		return
+	}
+
+	summary := TestResultSummary{}
+	err := m.Database.
+		Where(&TestResultSummary{BuildID: buildID, ArtifactID: artifactID}).
+		Find(&summary).
+		Error
+
+	if err != nil {
+		ginutil.WriteDBReadError(c, err, fmt.Sprintf(
+			"Failed fetching test result summary from test with ID %d for build with ID %d from database.",
+			artifactID, buildID))
+		return
+	}
+
+	c.JSON(http.StatusOK, summary)
+}
+
+// getBuildTestResultDetailsHandler godoc
+// @summary Get all test result details for specified test
+// @tags test-result
+// @param buildid path int true "Build ID"
+// @param artifactId path int true "Artifact ID"
+// @success 200 {object} []TestResultDetail
+// @failure 400 {object} problem.Response "Bad Request"
+// @failure 502 {object} problem.Response "Database is unreachable"
+// @router /build/{buildid}/test-result/summary/{artifactId}/detail [get]
+func (m buildTestResultModule) getBuildTestResultDetailsHandler(c *gin.Context) {
+	buildID, ok := ginutil.ParseParamUint(c, "buildid")
+	if !ok {
+		return
+	}
+
+	artifactID, ok := ginutil.ParseParamUint(c, "artifactId")
+	if !ok {
+		return
+	}
+
+	details := []TestResultDetail{}
+	err := m.Database.
+		Where(&TestResultDetail{BuildID: buildID, ArtifactID: artifactID}).
+		Find(&details).
+		Error
+
+	if err != nil {
+		ginutil.WriteDBReadError(c, err, fmt.Sprintf(
+			"Failed fetching test result details from test with ID %d for build with ID %d from database.",
+			artifactID, buildID))
+		return
+	}
+
+	c.JSON(http.StatusOK, details)
+}
+
+// getBuildTestResultListSummaryHandler godoc
+// @summary Get test result list summary of all tests for specified build
+// @tags test-result
+// @param buildid path int true "Build ID"
+// @success 200 {object} TestResultListSummary
+// @failure 400 {object} problem.Response "Bad Request"
+// @failure 502 {object} problem.Response "Database is unreachable"
+// @router /build/{buildid}/test-result/list-summary [get]
+func (m buildTestResultModule) getBuildTestResultListSummaryHandler(c *gin.Context) {
+	buildID, ok := ginutil.ParseParamUint(c, "buildid")
+	if !ok {
+		return
+	}
+
+	listSummary := TestResultListSummary{BuildID: buildID}
+
+	err := m.Database.
+		Model(&TestResultSummary{}).
+		Select("sum(failed) as Failed, sum(passed) as Passed, sum(skipped) as Skipped").
+		Where(&listSummary).
+		Scan(&listSummary).
+		Error
+
+	if err != nil {
+		ginutil.WriteDBReadError(c, err, fmt.Sprintf(
+			"Failed fetching test result summaries for build with ID %d from database.",
+			buildID))
+		return
+	}
+
+	listSummary.Total = listSummary.Failed + listSummary.Passed + listSummary.Skipped
+
+	c.JSON(http.StatusOK, listSummary)
+}
+
+type xmlInnerString struct {
+	InnerXML string `xml:",innerxml"`
+}
```

Comment:
Why not embed this type as well inside `trxTestRun`?

Reply:
I decided to keep it outside because it's used twice.